docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,266 @@
+
+_Note: In some situations, colocated types can still be exported if needed by other modules (e.g. React Component Props)_
+
+Example:
+
+```typescript
+// config.ts
+
+interface Config {
+  prop1: string;
+  prop2: number;
+}
+
+function update(config: Config): Config {
+  // ...
+}
+```
+
+```typescript
+// Box.tsx
+
+import * as React from 'react';
+
+export interface TaskProps {
+  name: string;
+  priority: 'high' | 'medium' | 'low';
+  due: Date;
+}
+
+export const TaskProps = ({ name, priority, due }: TaskProps) => {
+  // ...
+};
+```
+
+### b. Definitions Files
+
+**When types are shared across multiple files or modules, or when dealing with types that does not belong to a specific module**, it is recommended to use definition files (`.d.ts`) to separate type declarations from their usage.
+
+Definitions files provide a centralized location for types, making it easier to maintain and reuse them.
+
+Example:
+
+```typescript
+// service.d.ts
+interface CommonService {
+  [key: keyof any]: unknown;
+}
+
+// users.ts
+const usersService: CommonService = {
+  // ...
+};
+
+// books.ts
+const booksService: CommonService = {
+  // ...
+};
+```
+
+### c. Utils
+
+**For type utilities used across different packages**, it is mandatory to place them in their related module's definition file (.d.ts).
+
+- Strapi utils **must** be placed in `@strapi/strapi`'s types module
+- Packages' specific utils **must** be placed in their respective types' module
+
+Basically, utility modules contain reusable type definitions that can be imported wherever needed. This approach promotes code reusability and separation of concerns.
+
+Example:
+
+```typescript
+// core/uid.d.ts
+export type Admin = `admin::${string}`;
+
+// admin.ts
+import type { UID } from '@strapi/strapi';
+
+function getService(uid: UID.Admin): object {
+  // ...
+}
+```
+
+Choose the appropriate approach based on the size and complexity of your feature and the level of reusability of the types.
+
+## 2. Modules, Imports and Exports
+
+### a. Modules
+
+In TypeScript, modules can be used to organize and encapsulate related types and code. Modules provide a way to mimic namespaces and avoid polluting the global scope with numerous type declarations. By grouping related types within modules, it is possible to improve code organization and make it easier to navigate and understand the codebase.
+
+This is especially true when working with packages exports or utilities, where the exported components needs to be scoped.
+
+Example:
+
+```typescript
+// service.d.ts
+export type Admin = Record<string, any>;
+
+// uid.d.ts
+export type Admin = `admin::${string}`;
+
+// index.d.ts
+export * as Service from './service';
+export * as UID from './uid';
+
+// src/index.ts
+import type { Service, UID } from './index';
+
+type AdminService = Record<UID.Admin, Service.Admin>;
+```
+
+### b. Imports
+
+When importing **only types** from a module, `import type` must be used.
```

Comment:
Will this be enforce with eslint ? If so I would remove them from the guidelines so it only includes things we haven't a lint rule for :)

Reply:
What about keeping this + adding a section about eslint + adding backlinks between both?

Reply:
Also, 2 things:
- It's possible to enforce that with an eslint rule
- It's also possible to have a mix that preserves both: `import { type Article, createArticle } from 'smth';`

Reply:
Great news. If we ever change the rule the guidelines won't be up to date. It's really hard to maintain those kind of things that less is better IMO. I would instead add a section that link to the strapi commong eslint config and the local config file to know about all the others rules